Add hosted by archive addon module

ligangty opened the pull request:
Includes:
   * REST endpoint:
admin/stores/maven/hosted/{repo}/compressed-content?ignoreBeforeRoot=xxx
   * java client module
   * Config to disable/enable
   * Some ftests

jdcasey commented on addons/hosted-by-archive/jaxrs/src/main/java/org/commonjava/indy/hostedbyarc/bind/jaxrs/IndyHostedByArchiveResource.java:
```diff
@@ -0,0 +1,165 @@
+                                           final @Context HttpServletRequest request,
+                                           final @Context SecurityContext securityContext )
+    {
+        return createHostedByZip( name, uriInfo, ignoreBeforeRoot, form, request, securityContext );
```
I wonder if we can probe these for Zip compat here? I did some looking around using hexedit, and zip files all seem to have the same header: `0x50 0x4B 0x03 0x04` that we might be able to read to determine whether this is likely to be a zip file we can open.

I'd also like us to be able to consume the HttpServletRequest InputStream from an endpoint marked with @Consumes("application/zip") via POST / PUT, since this will be a little more machine-friendly than constructing a form post (I think).

jdcasey commented on addons/hosted-by-archive/client-java/src/main/java/org/commonjava/indy/hostedbyarc/client/IndyHostedByArchiveClientModule.java:
```diff
@@ -0,0 +1,119 @@
+        try
+        {
+            FileBody fileBody = new FileBody( zipFile, ContentType.APPLICATION_OCTET_STREAM );
+            HttpEntity entity = MultipartEntityBuilder.create()
```
If we create a REST endpoint to use `@Consumes("application/zip")` and the servlet input stream, this call would be simpler. This would also help us handle larger files.

Is there a reason you're using a multipart/form for this?

ligangty replied:
Not any special reasons. Will change it to streaming way.

ligangty commented on addons/hosted-by-archive/common/src/main/java/org/commonjava/indy/hostedbyarc/HostedByArchiveManager.java:
```diff
@@ -0,0 +1,312 @@
+        }
+        try
+        {
+            return Files.probeContentType( file.toPath() ).contains( "application/zip" );
```
Will do.

ligangty commented on addons/hosted-by-archive/client-java/src/main/java/org/commonjava/indy/hostedbyarc/client/IndyHostedByArchiveClientModule.java:
```diff
@@ -0,0 +1,119 @@
+
+    public static final String HOSTED_BY_ARC_PATH = "admin/stores/maven/hosted";
+
+    public HostedRepository createRepo( final File zipFile, final String repoName, final String ignoreBeforeRoot )
```
Will have a look to see how to do this.

ligangty commented on addons/hosted-by-archive/common/src/main/java/org/commonjava/indy/hostedbyarc/HostedByArchiveManager.java:
```diff
@@ -0,0 +1,312 @@
+                                              final String pathPrefix, final boolean deleteFilesAfterUnzip )
+            throws IndyWorkflowException
+    {
+        final File archive;
```
Not sure if we can wrap all events sending part in this single addon. Need to think about it.

ligangty commented on addons/hosted-by-archive/jaxrs/src/main/java/org/commonjava/indy/hostedbyarc/bind/jaxrs/FileUploadForm.java:
```diff
@@ -0,0 +1,44 @@
+
+    @FormParam( PARAM_FILE )
+    @PartType( "application/octet-stream" )
+    public void setData( byte[] data )
```
Will do. Maybe this file will be ignored.

ligangty commented:
@jdcasey In new commits, I've changed to use stream in REST endpoint, and I removed the usage of temp zip file and temp files for all zip entries, and directly store all zip entry in to created hosted repo.
But there is one potential issue for performance. With new way, we can not use multi-thread for storing the content because of the ZipInputStream(zis) non-sharing. All zip entries are wrapped in one single zis, and I don't find a good way to share this zis in multiple thread. So maybe this will bring some performance problem when storing huge zip files.

ligangty commented:
retest this please
